_studio/shared/src/libmfx_allocator_vaapi.cpp

```diff
@@ -110,6 +106,95 @@ unsigned int ConvertMfxFourccToVAFormat(mfxU32 fourcc)
+    attrib[attrCnt].value.type      = VAGenericValueTypeInteger;
+    attrib[attrCnt++].value.value.i = va_fourcc;
+
+    *format               = va_fourcc; // ???
+
+    switch (fourcc)
+    {
+    case MFX_FOURCC_VP8_NV12:
+            // special configuration for NV12 surf allocation for VP8 hybrid encoder is required
+            attrib[attrCnt].type            = VASurfaceAttribUsageHint;
+            attrib[attrCnt].flags           = VA_SURFACE_ATTRIB_SETTABLE;
+            attrib[attrCnt].value.type      = VAGenericValueTypeInteger;
+            attrib[attrCnt++].value.value.i = VA_SURFACE_ATTRIB_USAGE_HINT_ENCODER;
+            break;
+        case MFX_FOURCC_VP8_MBDATA:
+            // special configuration for MB data surf allocation for VP8 hybrid encoder is required
+            attrib[0].value.value.i = VA_FOURCC_P208;
+            *format                  = VA_FOURCC_P208;
+            break;
+        case MFX_FOURCC_NV12:
+            *format = VA_RT_FORMAT_YUV420;
+            break;
+        case MFX_FOURCC_UYVY:
+        case MFX_FOURCC_YUY2:
+            *format = VA_RT_FORMAT_YUV422;
+            break;
+            break;
+        case MFX_FOURCC_A2RGB10:
+            *format = VA_RT_FORMAT_RGB32_10BPP;
+            break;
+        case MFX_FOURCC_RGBP:
+            *format = VA_RT_FORMAT_RGBP;
+            break;
+        case MFX_FOURCC_RGB4:
+        case MFX_FOURCC_BGR4:
+            *format = VA_RT_FORMAT_RGB32;
+            /*
+             *  Enable this hint as required for creating RGB32 surface for MJPEG.
+             */
+            if ((memType & MFX_MEMTYPE_VIDEO_MEMORY_ENCODER_TARGET)
+                    && (memType & MFX_MEMTYPE_FROM_ENCODE))
+            {
+              //  Input Attribute Usage Hint
+                attrib[attrCnt].flags           = VA_SURFACE_ATTRIB_SETTABLE;
+                attrib[attrCnt].type            = VASurfaceAttribUsageHint;
+                attrib[attrCnt].value.type      = VAGenericValueTypeInteger;
+                attrib[attrCnt++].value.value.i = VA_SURFACE_ATTRIB_USAGE_HINT_ENCODER;
+            }
+            break;
+        default:
+            break;
+    }
+
+    return attrCnt;
+}
+
+static bool isFourCCSupported(unsigned int va_fourcc)
+{
+    switch (va_fourcc)
+    {
+        case VA_FOURCC_NV12:
+        case VA_FOURCC_YV12:
+        case VA_FOURCC_YUY2:
+        case VA_FOURCC_ARGB:
+        case VA_FOURCC_ABGR:
+#ifdef MFX_ENABLE_RGBP
+        case VA_FOURCC_RGBP:
+#endif
+        case VA_FOURCC_UYVY:
+        case VA_FOURCC_P208:
+        case VA_FOURCC_P010:
+        case VA_FOURCC_AYUV:
+        case VA_FOURCC_R5G6B5:
+#if (MFX_VERSION >= 1027)
+        case VA_FOURCC_Y210:
+        case VA_FOURCC_Y410:
+#endif
+            return true;
+        default:
+            return false;
+    }
+}
+
 static mfxStatus ReallocImpl(mfxDefaultAllocatorVAAPI::mfxWideHWFrameAllocator* pSelf, vaapiMemIdInt *vaapi_mid, mfxFrameSurface1 *surf)
```

Comment:
The only thing u need from `mfxWideHWFrameAllocator` is VA Display. It makes sense to pass display itself as argument rather then allocator.